### DISCUSSION TITLE
Modifying Website Title

### DISCUSSION THREAD

**Discussion by dm544** [vradarserver/vrs#93] [Q&A] [answered]
Hello,

I'd like to modify the website's title from `Virtual Radar (countAircraft)` to something like `(countMilAircraft) Virtual Radar (countAircraft)` where countAircraft is the number of aircraft in the list and countMilAircraft is the number of military aircraft on the list.

I've found in titleUpdater.cs where the title is updated, but if I try to change the values, the website breaks (which is what the Custom Content Plugin says will happen, so that's fair). I've been tinkering around with the Custom Content Plugin, trying to get any change to go into effect, but haven't gotten anything to work.

Any help would be appreciated!

**Comment by vradarserver** [comment, marked as answer]
Unfortunately I didn't put a whole lot of thought into how scripts might take over the formatting of the page's title. However, you can replace any of the site's files with your own versions if you want to - you just need to go through the custom content plugin to do it.

1. Create a folder somewhere on your machine. For this example I'll assume `c:\vrs-custom`
2. In that folder create another called `script`
3. In script create another folder called `vrs`
4. Copy `titleUpdater.js` (that's the `.js` file, not the .ts file that it was transpiled from) into the `vrs` folder (e.g. `c:\vrs-custom\script\vrs`).
5. In the Custom Content Plugin set the **Site root folder** entry to the folder you created in step 1 (e.g. `c:\vrs-custom`).

You might need to restart the server depending on whether you have bundling switched on, I can't remember... maybe restart it just in case :)

Any files that are in your custom folder will be served instead of the original, so you should ideally only have custom versions in there - i.e. don't copy the entire site across to your custom folder. Just copy the one JS file that you want to change.

In general you should try to avoid replacing entire JavaScript files if you can help it. But with that said, the title updater doesn't change very often and it's very small, so it shouldn't be too difficult to tell whether you need to take a new version of it after an update.

**Reply by dm544** [reply]
Awesome, that did the trick, thanks! The plugin still requires a .html, but I just put a blank template in and it works. Now I need to figure out how to count the number of military marked aircraft on the list. If you happen to know how to do that off the top of your head, I'd appreciate it, but I'll start trying to figure that out myself.

**Reply by vradarserver** [reply]
The plugin doesn't need to add a script, you can just use the custom content plugin to set a custom root. Perhaps you already had a custom script set up and its content was wrong? If so then you can select and then delete it if it's not doing anything.

Regarding changing the script to add a count of military aircraft - I'm assuming that you can muddle along and broadly see how the JavaScript is laid out?

First of all, there's a function at the start called TitleUpdater() that looks like this, except it doesn't have that `this._PreviousMilitaryCount = -1;` line in it. Add the line, or copy and paste this over the existing function:

```
        function TitleUpdater() {
            this._AircraftList = null;
            this._AircraftListUpdatedHookResult = null;
            this._AircraftListPreviousCount = -1;
this._PreviousMilitaryCount = -1;            
            this._LocaleChangedHookResult = VRS.globalisation.hookLocaleChanged(this.localeChanged, this);
        }
```

I left out the indent for the line so that it would stand out. Ordinarily you'd want to indent everything, but given you'll need to repeat these changes in the future if the module changes then maybe it's a good idea to keep your modifications standing out.

The other function you need to change is `refreshAircraftListCount` - same thing, either copy and paste this on top of the original version (taking care not to overwrite any of the functions above or below it) or otherwise adjust it so that it looks like this:

```
        TitleUpdater.prototype.refreshAircraftCount = function (forceRefresh) {
            if (this._AircraftList) {
                var count = this._AircraftList.getCountTrackedAircraft();
var countMilitary = 0;
this._AircraftList.foreachAircraft(function(ac) { if(ac.isMilitary.val) ++countMilitary; });
//                if (forceRefresh || count !== this._AircraftListPreviousCount) {
if (forceRefresh || count !== this._AircraftListPreviousCount || countMilitary != this._PreviousMilitaryCount) {
//                    document.title = VRS.$$.VirtualRadar + ' (' + count + ')';
document.title = '(' + countMilitary + ') ' + VRS.$$.VirtualRadar + ' (' + count + ')';
this._PreviousMilitaryCount = countMilitary;
                    this._AircraftListPreviousCount = count;
                }
            }
        };
```

I think that should do it.

I've not tested it against military aircraft, there aren't any in my feed at the moment. If things need adjusting then the line that does the counting is this one:

```
this._AircraftList.foreachAircraft(function(ac) { if(ac.isMilitary.val) ++countMilitary; });
```

`this._AircraftList` is a reference to the aircraft list, `foreachAircraft` is a function that loops through the aircraft list and calls a function for each entry, then the function that's supplied accepts the aicraft as a variable called `ac` and increments a counter if `isMilitary.val` is set.

The bit that updates the title if the counters have changed is

```
document.title = '(' + countMilitary + ') ' + VRS.$$.VirtualRadar + ' (' + count + ')';
```

`VRS.$$.VirtualRadar` is a translatable string, it's Virtual Radar in English.

**Reply by dm544** [reply]
This is perfect, thank you so much!

I added an if statement to only show the countMilitary if it's not 0
```
if (countMilitary == 0) {
    document.title = VRS.$$.VirtualRadar + ' (' + count + ')';
} else {
    document.title = '(' + countMilitary + ') ' + VRS.$$.VirtualRadar + ' (' + count + ')';
}
```

I also added a count for aircraft with a userTag
`this._AircraftList.foreachAircraft(function (ac) { if (ac.userTag.val !== undefined) ++countMilitary; });`

Still waiting for a mil aircraft to come into scope, but it works as intended for any aircraft with a userTag.
Thank you again!

**Reply by vradarserver** [reply]
One minor point - your last statement either needs a new counter besides `countMilitary` or you need to increment it if either the user tag is defined or the aircraft is military - otherwise an aircraft that is both military and has a user tag will be counted twice. I think it should be something like:

```
this._AircraftList.foreachAircraft(function(ac) { if(ac.isMilitary.val || ac.userTag.val !== undefined) ++countMilitary; });
```

**Reply by dm544** [reply]
Haha yep, I thought of that earlier this morning. Changed it to the line you posted exactly.
Thanks!
